doc/architectural/compilation-and-development.md

```diff
@@ -3,43 +3,106 @@
+Instructions for compiling CBMC using makefiles are
+available in
+[COMPILING.md](https://github.com/diffblue/cbmc/blob/develop/COMPILING.md#what-architecture)
+in the root of the CBMC repository. They cover linux, Solaris 11,
+FreeBSD 11, MacOS X and Windows.
 
-To be documented.
 
-## Running tests ##
+\subsection compilation-and-development-subsection-cmake-files CMake files
 
-### Regression tests ###
+There is experimental support for compiling using CMake. Instructions are
+available in
+[COMPILING.md](https://github.com/diffblue/cbmc/blob/develop/COMPILING.md#working-with-cmake-experimental)
+in the root of the CBMC repository.
 
-The regression tests are contained in the `regression/` folder.
-They are grouped into directories for each of the tools/modules.
-Each of these contains multiple directories, each of which contains
-input files and one or more `.desc` files that describe what command
-to run, what output is expected and so on. There is a Perl script,
-`test.pl` that is used to invoke the tests as:
+
+\subsection compilation-and-development-subsection-personal-configuration Personal configuration
+
+\subsubsection compilation-and-development-subsubsection-config-inc config.inc
+
+Two files,
+[config.inc](https://github.com/diffblue/cbmc/blob/develop/src/config.inc) and
+[common](https://github.com/diffblue/cbmc/blob/develop/src/common), are
+included in every makefile. 
+[config.inc](https://github.com/diffblue/cbmc/blob/develop/src/config.inc)
+contains configuration options
+relating to compilation so that they can be conveniently edited in one place.
+[common](https://github.com/diffblue/cbmc/blob/develop/src/common)
+contains commands that are needed in every makefile but which the
+developer is not expected to edit. (There is also
+[another config.inc](https://github.com/diffblue/cbmc/blob/develop/jbmc/src/config.inc)
+which is also included in every makefile in the `jbmc` folder.) 
+ 
+
+\subsubsection compilation-and-development-subsubsection-macro-debug Macro DEBUG
+
+If the macro `DEBUG` is defined during compilation of CBMC (for example by
+using a compiler flag) then extra debug code will be included. This includes
+print statements and code checking that data structures are as expected.
+
+
+\section compilation-and-development-section-running-tests Running tests
+
+\subsection compilation-and-development-subsection-regression-tests Regression tests
+
+The regression tests are contained in the `regression/` folder. Inside
+`regression/` there is a directory for each of the tools/modules. Each of
+these contains multiple test directories, with names describing
+what they test. When there are multiple tests in a test directory then
+they should all be testing the same thing. Each test directory contains
+input files and one or more test description files,
+which have the ending `.desc`. The test description files describe what command
+to run, what output is expected and so on. The test framework is an in-house
```

Comment:
:pick: Remove `in-house`. Most of the code is "in-house" (and that part is not one we're particularly proud of).

Reply:
Would you accept "hand-rolled"?

Reply:
Fine